### PR TITLE
Drop `{{ pin_compatible('numpy', max_pin='x') }}`

### PR DESCRIPTION
**Pull request by jakirkham**
## Description

Part of issue: https://github.com/rapidsai/build-planning/issues/82

Drop `{{ pin_compatible('numpy', max_pin='x') }}` as it is no longer needed. `numpy` has its own `run_exports`, which constraints `numpy` to an API compatible version. More details in issue: https://github.com/orgs/rapidsai/projects/132

So `cudf` now uses that in its recipe builds. Also update `requirements/run` to set the `numpy` lower bound to `1.23` as required by us.

Lastly add todo comments for NumPy 2 update lines.

## Checklist
- [ ] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [ ] New or existing tests cover these changes.
- [ ] The documentation is up to date with these changes.

### CODE REVIEW

**Review by jameslamb — APPROVED** [review]
Makes sense to me, I double-checked and right now `numpy`'s run export is `>=1.19,<3`.

https://github.com/conda-forge/numpy-feedstock/blob/10b2fa510901ba22db1a00b81e5aa4c646a11b03/recipe/meta.yaml#L6

https://github.com/conda-forge/numpy-feedstock/blob/10b2fa510901ba22db1a00b81e5aa4c646a11b03/recipe/meta.yaml#L34-L35

File: conda/recipes/cudf/meta.yaml [hunk 1]
```diff
@@ -64,6 +64,7 @@ requirements:
     - rapids-build-backend >=0.3.0,<0.4.0.dev0
     - scikit-build-core >=0.7.0
     - dlpack >=0.8,<1.0
+    # TODO: Change to `2.0` for NumPy 2
     - numpy 1.23
```
Comment (jakirkham) [inline conda/recipes/cudf/meta.yaml:68]: Once we update to NumPy 2, that will be true

Right now we build with NumPy 1.23, which means [`run_exports` adds `{{ pin_subpackage("numpy") }}`]( https://github.com/conda-forge/numpy-feedstock/blob/ddbe092d07a0a6f57c3fd6854d7c17062c8fa784/recipe/meta.yaml#L26-L27 ). The [defaults of `pin_subpackage`]( https://docs.conda.io/projects/conda-build/en/stable/resources/variants.html#extra-jinja2-functions ) are to tightly pin on the lower bound and pin the upper bound based on major version. So this turns into `numpy >=1.23.5,<2.0a0`

Comment (jameslamb) [reply conda/recipes/cudf/meta.yaml:68]: oh got it, right. Ok sounds good to me.

File: conda/recipes/cudf/meta.yaml [hunk 2]
```diff
@@ -82,7 +83,8 @@ requirements:
     - pandas >=2.0,<2.2.3dev0
     - cupy >=12.0.0
     - numba >=0.57
-    - {{ pin_compatible('numpy', max_pin='x') }}
+    # TODO: Update `numpy` in `host` when dropping `<2.0a0`
+    - numpy >=1.23,<2.0a0
```
Comment (jakirkham) [inline conda/recipes/cudf/meta.yaml:87]: While this seems duplicative to what `numpy`'s `run_exports` does, this same pinning is used throughout RAPIDS. Also it gives a foothold for RDFG changes to be made to move to NumPy 2. When that does happen, we have a comment above to remind us to make other needed changes in this recipe

**Comment by jakirkham** [comment]
Thanks James! 🙏

**Pull request merged**